arcticbio: Update puzzles.py

freckletonj opened the pull request:
1. Modified Magnets combinator to handle undefined block_positions metadata more gracefully.
2. Extended gen_some_blocks to include block_positions metadata (could use more validation)
3. Added extension array with additional puzzle definitions

arcticbio committed:
Update puzzles.py

1. Modified Magnets combinator to handle undefined block_positions metadata more gracefully.
2. Extended gen_some_blocks to include block_positions metadata (could use more validation)
3. Added extension array with additional puzzle definitions

freckletonj commented:
As of right now some of these puzzles don't quite "work", so, i'm gonna go through and triage a bit, for instance:

```
magnets-5 : output seems to be dropping blocks, eg the last row
invert-colors + move-to-pivot: seems weird, can't put my finger on it
translate(3) + right-align: the translation seems irrelevant since they get right-aligned anyway
```

![image](https://github.com/user-attachments/assets/98e04074-ecdd-4411-9287-95e12a78584d)

freckletonj committed:
add `arcticbio` to #Gratitude

arcticbio committed:
Constrain "magnets"

As written, Magnets will only work with exactly two blocks, since any third block could shift on top of another block and obfuscate the learnable translation.  It could be rewritten to select two random adjacent blocks.  For now, I will constrain the function and comment the limitation for clarity

arcticbio committed:
Magnets work and demo puzzles

Additional handling on magnets function to handle more metadata conditions - it may be worthwhile to harmonize the initializations on the metadata front?

Created two demo puzzles to illustrate conflicts which will occur when adjusting the size or position of blocks in the current composition architecture.

arcticbio commented:
Without significant rewrite, Magnets will only provide meaningful results with two blocks; any more than two, one runs the risk of the shifted blocks layering on top of another blocks, obfuscating the pattern.  Could potentially be rewritten to select two adjacent blocks and continue the translation.  For now I will comment magnets with the constraint and adjust the puzzle.

I also added color 5 back into the base list, since it appears that we are explicitly removing it from use cases where the value is used for pivot.

Overall, I'm struggling with the limited composability of scaling and shifting functions, since they will frequently overwrite each other or run off the working space; two example puzzles added to this branch to explore the outcome and determine how they should be handled.

arcticbio committed:
Condition handling, block metadata consistency, improved composability

Added block_position metadata across initialization functions
Built block deconfliction mechanism for improved composability
Extended magnets to allow for 'n' blocks for greater variability in training data without additional puzzle definitions
Extended rotate_block_pixels to allow for randomization of shift value for greater variability in training data without additional puzzle definitions
Added a couple of puzzles to demonstrate the extended capability

arcticbio commented:
Improved composability and edge case handling to allow for more of the puzzles to work with higher degree of consistency without human intervention / oversight.

There are a few new puzzles in place for general review, and I will now go back to the original list, review through the lens of these more robust translation and initialization functions, and bring back in those which are now more transparent and consistent.

freckletonj commented on arc_like/puzzles.py:
```diff
@@ -34,6 +35,15 @@ class Sequence:
 # clean block becomes the input, so we `swap` them after noising.
 Combinator = Callable[[Sequence], Sequence]
 
+def find_all_contiguous_starts(available_positions, size):
```
:fire: awesome idea

freckletonj commented:
Thank you @arcticbio , this is looking really awesome!
I'll spend some time tomorrow going through this code too. I really like the `contiguous blocks` idea :heart:

arcticbio committed:
Worked on wrap consistency and composition standards

 - Magnets extension to allow for reversal of large and small block anchor role
 - Added wrapping support to magnets and rotation
 - Created additional puzzles which leverage the extended composability

arcticbio committed:
Support for random length in single block generator

 - Added optional "length" parameter on gen_one_block to allow for greater puzzle variability (undefined randomizes the value)
 - Added handling for blocks which span to end of sequence length in extend to pivot and move to pivot (resolved Stop Iteration error where blocks bump against sequence end)

arcticbio committed:
Added final puzzle

Floating length single block with color inversion

arcticbio commented:
Hopefully this give you a good starting point for the weekend and maybe a good point to merge the branch.  Puzzle variety is expanded, and known error conditions are resolved.  

I ran all puzzles with 50 samples each over several iterations with floating seed to allow for initial load testing, and there were no errors; spot check review did not indicate any glaring output errors.

freckletonj committed:
Merge branch 'master' into arcticbio-proposed-puzzles

freckletonj committed:
tweaks

freckletonj commented:
@arcticbio Thank you!  I just fixed up some merge conflicts, but other than that this looks awesome!
I like that you robustified it too, Merging!
